src/test/lib/optimizer/strategy/null_scan_removal_rule_test.cpp

```diff
@@ -0,0 +1,103 @@
+#include <memory>
+
+#include "expression/expression_functional.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+#include "optimizer/strategy/null_scan_removal_rule.hpp"
+#include "strategy_base_test.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+class NullScanRemovalRuleTest : public StrategyBaseTest {
+ public:
+  void SetUp() override {
+    rule = std::make_shared<NullScanRemovalRule>();
+    mock_node = MockNode::make(MockNode::ColumnDefinitions{{DataType::Int, "a"}, {DataType::Float, "b"}});
+    mock_node_column = mock_node->get_column("a");
+
+    Hyrise::get().storage_manager.add_table("nullable_table",
+                                            load_table("resources/test_data/tbl/int_float_null_1.tbl", 2));
+    Hyrise::get().storage_manager.add_table("table", load_table("resources/test_data/tbl/int_float4_or_1.tbl", 2));
+    nullable_table_node = StoredTableNode::make("nullable_table");
+    table_node = StoredTableNode::make("table");
+    nullable_table_node_column = lqp_column_(nullable_table_node, ColumnID{0});
+    table_node_column = lqp_column_(table_node, ColumnID{0});
+  }
+  std::shared_ptr<MockNode> mock_node;
+  std::shared_ptr<NullScanRemovalRule> rule;
+  std::shared_ptr<LQPColumnExpression> mock_node_column, nullable_table_node_column, table_node_column;
+  std::shared_ptr<StoredTableNode> nullable_table_node, table_node;
+};
+
+TEST_F(NullScanRemovalRuleTest, LQPNodeTypeIsNotPredicate) {
+  // The rule can't apply on a node that is not of type Predicate.
+  const auto actual_lqp = apply_rule(rule, mock_node);
+  const auto expected_lqp = mock_node->deep_copy();
+
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+TEST_F(NullScanRemovalRuleTest, PredicateIsNotNullExpression) {
+  // The rule can't apply on a predicate that is not a `null expression`.
+  const auto input_lqp = PredicateNode::make(equals_(mock_node_column, 42), mock_node);
+  const auto actual_lqp = apply_rule(rule, input_lqp);
+  const auto expected_lqp = input_lqp->deep_copy();
```

Comment:
The input_lqp is modified as part of `apply_rule`. If you deep copy it after you apply the rule, I believe that you are not testing anything. Did you see this pattern somewhere else? If yes, we should look into that.

Reply:
That's totally my fault 🙈 . I changed the code to execute the deep copy before applying the rule.

Reply:
I don't yet see a test where the LQP is modified. If `expected_lqp` is a deep copy of the `input_lqp` and you assert that both are equal, it means that the rule did not do anything. That is ok for the edge cases, but unless I am blind I don't see a test that covers the happy case.

Reply:
Assuming I understood `PredicateNode::make` (enable_make_for_lqp_node.hpp) correctly, then in `TEST_F(NullScanRemovalRuleTest, TableColumnDefinitionIsNotNullable)` we modify the LQP. We have the following LQP bevore applying the node: 

```
[[StoredTable]] -> [[Predicate] note IS NOT NULL]
```

After applying the rule we should only have: 

```
 [[StoredTable]]
```

Reply:
Ah, got it. I just saw the `deep_copy` and didn't read the left side.